index.js

```diff
@@ -39,116 +33,113 @@ function fromGeojsonVt (layers) {
   return fromVectorTileJs({layers: l})
 }
 
-/**
- * Prepare the given layer to be serialized by the auto-generated pbf
- * serializer by encoding the feature geometry and properties.
- */
-function prepareLayer (layer) {
-  var preparedLayer = {
-    name: layer.name || '',
-    version: layer.version || 1,
-    extent: layer.extent || 4096,
-    keys: [],
-    values: [],
-    features: []
+function command (cmd, length) {
+  return (length << 3) + (cmd & 0x7)
+}
+
+function zigzag (num) {
+  return (num << 1) ^ (num >> 31)
+}
+
+function writeTile (tile, pbf) {
+  for (var key in tile.layers) {
+    pbf.writeMessage(3, writeLayer, tile.layers[key])
   }
+}
 
+function writeLayer (layer, pbf) {
+  pbf.writeVarintField(15, layer.version || 1)
+  pbf.writeStringField(1, layer.name || '')
+  pbf.writeVarintField(5, layer.extent || 4096)
+
+  var i
+  var keys = []
+  var values = []
   var keycache = {}
   var valuecache = {}
 
-  for (var i = 0; i < layer.length; i++) {
-    var feature = layer.feature(i)
-    feature.geometry = encodeGeometry(feature.loadGeometry())
-
-    var tags = []
-    for (var key in feature.properties) {
-      var keyIndex = keycache[key]
-      if (typeof keyIndex === 'undefined') {
-        preparedLayer.keys.push(key)
-        keyIndex = preparedLayer.keys.length - 1
-        keycache[key] = keyIndex
-      }
-      var value = wrapValue(feature.properties[key])
-      var valueIndex = valuecache[value.key]
-      if (typeof valueIndex === 'undefined') {
-        preparedLayer.values.push(value)
-        valueIndex = preparedLayer.values.length - 1
-        valuecache[value.key] = valueIndex
-      }
-      tags.push(keyIndex)
-      tags.push(valueIndex)
+  function writeFeature (feature, pbf) {
+    if (feature.id !== undefined) {
+      pbf.writeVarintField(1, feature.id)
     }
 
-    feature.tags = tags
-    preparedLayer.features.push(feature)
+    pbf.writeMessage(2, function () {
```

Comment:
You can avoid using closures here (which have a certain overhead) if you take advantage of the third argument to `writeMessage`. This may be the root of the simple features regression.